browser/brave_ads/search_result_ad/search_result_ad_handler.cc

```diff
@@ -0,0 +1,305 @@
+
+    if (!property->values->is_entity_values() ||
+        property->values->get_entity_values().empty()) {
+      LOG(ERROR) << "Search result ad attributes list is empty";
+      return SearchResultAdsList();
+    }
+
+    for (const auto& ad_entity : property->values->get_entity_values()) {
+      if (!ad_entity || ad_entity->type != kSearchResultAdType) {
+        LOG(ERROR) << "Wrong search result ad type specified: "
+                   << ad_entity->type;
+        return SearchResultAdsList();
+      }
+
+      if (property->name == kTypePropertyName) {
+        continue;
+      }
+
+      SearchResultAdPtr search_result_ad = SearchResultAd::New();
+      search_result_ad->conversion = Conversion::New();
+
+      base::flat_set<base::StringPiece> found_attributes;
+      for (const auto& ad_property : ad_entity->properties) {
+        // Wrong attribute name
+        const auto* it = std::find(kSearchResultAdAttributes.begin(),
+                                   kSearchResultAdAttributes.end(),
+                                   base::StringPiece(ad_property->name));
+
+        if (it == kSearchResultAdAttributes.end()) {
+          LOG(ERROR) << "Wrong search result ad attribute specified: "
+                     << ad_property->name;
+          return SearchResultAdsList();
+        }
+        found_attributes.insert(*it);
+
+        if (!SetSearchAdProperty(ad_property, search_result_ad.get())) {
+          LOG(ERROR) << "Cannot read search result ad attribute value: "
+                     << ad_property->name;
+          return SearchResultAdsList();
+        }
+      }
+
+      // Not all of attributes were specified.
+      if (found_attributes.size() != kSearchResultAdAttributes.size()) {
+        std::vector<base::StringPiece> absent_attributes;
+        std::set_difference(kSearchResultAdAttributes.begin(),
+                            kSearchResultAdAttributes.end(),
+                            found_attributes.begin(), found_attributes.end(),
+                            std::back_inserter(absent_attributes));
+
+        LOG(ERROR) << "Some of search result ad attributes were not specified: "
+                   << base::JoinString(absent_attributes, ", ");
+
+        return SearchResultAdsList();
+      }
+
+      search_result_ads_list.push_back(std::move(search_result_ad));
+    }
+
+    // Creatives has been parsed.
+    break;
+  }
+
+  return search_result_ads_list;
+}
+
+SearchResultAdsList ParseWebPageEntities(blink::mojom::WebPagePtr web_page) {
+  for (const auto& entity : web_page->entities) {
+    if (entity->type != kSearchResultAdsListType) {
+      continue;
+    }
+
+    absl::optional<SearchResultAdsList> search_result_ads_list =
+        ParseSearchResultAdsListEntityProperties(entity);
+
+    if (search_result_ads_list) {
+      return std::move(*search_result_ads_list);
+    }
+  }
+
+  return SearchResultAdsList();
+}
+
+}  // namespace
+
+SearchResultAdHandler::SearchResultAdHandler() = default;
```

Comment:
so thinking about this more I think this should be a new keyed service in components/brave_ads so we keep everything encapsulated within ads and also we aren't scoping the lifetime of this handler to the tab helper. Since it's doing async processing (getting the metadata), the tab could be closed before it receives it and then it's lost.

Reply:
Great idea, thanks! There is another option here: 
We already have `AdsService` as a keyed service. So we could add `SearchResultAdHandler` as a member there, so its lifetime scope will be the same as AdsService. 
WDYT about this option?

Reply:
I have added `search_result_ad_aliases.h` because it now contains `SearchResultAdsList` type which is used in two headers: `search_result_ad_handler.h` and `parse_web_page_entities.h`